pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,186 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pleg
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	internalapi "k8s.io/cri-api/pkg/apis"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+	"k8s.io/klog/v2"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/utils/clock"
+)
+
+type EventedPLEG struct {
+	// The container runtime.
+	runtime kubecontainer.Runtime
+
+	runtimeService internalapi.RuntimeService
+
+	// The channel from which the subscriber listens events.
+	eventChannel chan *PodLifecycleEvent
+
+	// Cache for storing the runtime states required for syncing pods.
+	cache kubecontainer.Cache
+	// For testability.
+	clock clock.Clock
+}
+
+// NewGenericPLEG instantiates a new GenericPLEG object and return it.
+func NewEventedPLEG(runtime kubecontainer.Runtime, runtimeService internalapi.RuntimeService, eventChannel chan *PodLifecycleEvent,
+	cache kubecontainer.Cache, clock clock.Clock) PodLifecycleEventGenerator {
+	return &EventedPLEG{
+		runtime:        runtime,
+		runtimeService: runtimeService,
+		eventChannel:   eventChannel,
+		cache:          cache,
+		clock:          clock,
+	}
+}
+
+// Watch returns a channel from which the subscriber can receive PodLifecycleEvent
+// events.
+// TODO: support multiple subscribers.
+func (e *EventedPLEG) Watch() chan *PodLifecycleEvent {
+	return e.eventChannel
+}
+
+// Start spawns a goroutine to relist periodically.
+func (e *EventedPLEG) Start() {
+	go wait.Forever(e.watchEventsChannel, 0)
+}
+
+// Healthy check if PLEG work properly.
+func (e *EventedPLEG) Healthy() (bool, error) {
+	// GenericPLEG is declared unhealthy when relisting time is more than
+	// the relistThreshold. In case, EventedPLEG is turned on the value of
+	// relistingPeriod, relistingThreshold is adjusted to higher values and
+	// so health of Generic PLEG should take care of the condition checking
+	// higher relistThreshold between last two relist.
+
+	// EventedPLEG is declared unhealthy only if eventChannel is out of capacity.
+	if len(e.eventChannel) >= cap(e.eventChannel) {
+		return false, fmt.Errorf("pleg event channel capacity is full with %v events", len(e.eventChannel))
+	}
+	return true, nil
+}
+
+func (e *EventedPLEG) watchEventsChannel() {
+
+	containerEventsResponseCh := make(chan *runtimeapi.ContainerEventResponse, cap(e.eventChannel))
+	defer close(containerEventsResponseCh)
+
+	// Get the container events from the runtime.
+	go e.runtimeService.GetContainerEvents(containerEventsResponseCh)
+
+	e.processCRIEvents(containerEventsResponseCh)
+}
+
+func (e *EventedPLEG) processCRIEvents(containerEventsResponseCh chan *runtimeapi.ContainerEventResponse) {
+	for event := range containerEventsResponseCh {
+		e.updatePodStatus(event)
+		switch event.ContainerEventType {
```

Comment:
This switch could be made into a function to aid in testing. It could also be spawned as a go routine to asynchronously loop through the events. Currently, a write to the eventChannel could block. Thoughts?

Reply:
Actually, I had some follow-up code for this one:

```go

func (e *EventedPLEG) sendEvent(event *PodLifecycleEvent) {
	select {
	case e.eventChannel <- event:
	default:
		// record how many events were discarded due to channel out of capacity
		metrics.PLEGDiscardEvents.Inc()
		klog.ErrorS(nil, "Evented PLEG: Event channel is full, discarded pod lifecycle event")
	}
}
```

with something like: `e.sendEvent(&PodLifecycleEvent{ID: types.UID(event.PodSandboxMetadata.Uid), Type: ContainerDied, Data: event.ContainerId})` instead of current: `e.eventChannel <- &PodLifecycleEvent{ID: types.UID(event.PodSandboxMetadata.Uid), Type: ContainerDied, Data: event.ContainerId}`

Probably, we can add on top it with and make it either bit more modular and or more asynchronous with go routine(s) as said?

Reply:
Why do we need to make processCRIEvent async? Wouldn't that make event processing potentially out-of-order? (e.g. runtime sends Container create, then start, then stop. But kubelet process stop first, then start and then create)

Reply:
I agree with @ruiwen-zhao.  

@swghosh @rphillips I am not sure if making processCRIEvent async is a good idea. We should process the events in the same order as we receive from the runtime, even if that means waiting for the channel to clear up.